In [2]:
# Libraries import karo
import pandas as pd
import numpy as np
import yfinance as yf
import matplotlib.pyplot as plt
import seaborn as sns

# Data load karo
ticker = yf.Ticker("RELIANCE.NS")
df = ticker.history(start="2020-01-01", end="2026-04-05")

# Date clean karo
df.index = df.index.tz_localize(None)

# Unnecessary columns drop karo
df = df.drop(columns=["Dividends", "Stock Splits"])

print("Data loaded successfully! ✅")
print("Shape:", df.shape)

Data loaded successfully! ✅
Shape: (1549, 5)


In [3]:
# Saal ke hisaab se average closing price
df['Year'] = df.index.year
yearly_avg = df.groupby('Year')['Close'].mean().round(2)

print("Yearly Average Close Price:")
print(yearly_avg)

Yearly Average Close Price:
Year
2020     787.88
2021     999.51
2022    1152.05
2023    1153.91
2024    1414.42
2025    1382.70
2026    1424.83
Name: Close, dtype: float64


💡 Business Insight #1:

2020 se 2024 tak Reliance ki price almost 2x ho gayi! — ₹787 se ₹1414 tak! Ye ek investor ke liye bahut strong return hai. Ye insight tumhare project mein mention karna recruiter ko impress karega! 😎

In [4]:
# Month ke hisaab se average price
df['Month'] = df.index.month
monthly_avg = df.groupby('Month')['Close'].mean().round(2)

print("Monthly Average Close Price:")
print(monthly_avg)

Monthly Average Close Price:
Month
1     1119.18
2     1124.12
3     1084.80
4     1091.25
5     1123.91
6     1153.69
7     1198.87
8     1202.98
9     1216.93
10    1205.67
11    1192.11
12    1201.48
Name: Close, dtype: float64


💡 Business Insight #2:

March-April mein price historically kam hoti hai aur July-September mein sabse zyada! — iska matlab agar koi investor March mein buy kare aur September mein sell kare toh historically achha return milta hai! Ye ek real trading insight hai! 😎

In [5]:
# Detailed statistics
print("Close Price Statistics:")
print("=======================")
print(f"Sabse kam price:  ₹{df['Close'].min():.2f}")
print(f"Sabse zyada price: ₹{df['Close'].max():.2f}")
print(f"Average price:    ₹{df['Close'].mean():.2f}")
print(f"Median price:     ₹{df['Close'].median():.2f}")
print(f"Price ka std dev: ₹{df['Close'].std():.2f}")

Close Price Statistics:
Sabse kam price:  ₹395.48
Sabse zyada price: ₹1592.30
Average price:    ₹1158.57
Median price:     ₹1168.81
Price ka std dev: ₹242.06


💡 Mean vs Median — Samjhte Hain:

Mean = ₹1158 aur Median = ₹1168 — dono almost same hain!
Iska matlab data mein koi bahut bada outlier nahi hai — data balanced hai! ✅
Agar mean aur median mein bahut fark hota toh koi extreme value hoti data mein!

💡 Business Insight #3:

₹395 se ₹1592 tak — matlab Reliance ne 6 saalon mein 4x return diya! Ye ek bahut strong long-term investment rahi hai! 😎

In [6]:
# Volume statistics
print("Volume Statistics:")
print("==================")
print(f"Average daily volume: {df['Volume'].mean():.0f} shares")
print(f"Max volume day:       {df['Volume'].max()} shares")
print(f"Min volume day:       {df['Volume'].min()} shares")

# Sabse zyada volume wala din
max_vol_date = df['Volume'].idxmax()
print(f"\nSabse zyada trading hui: {max_vol_date.date()}")
print(f"Us din ka Close price:  ₹{df.loc[max_vol_date, 'Close']:.2f}")

Volume Statistics:
Average daily volume: 17996415 shares
Max volume day:       142683366 shares
Min volume day:       0 shares

Sabse zyada trading hui: 2020-04-22
Us din ka Close price:  ₹610.01


💡 Business Insight #4:

22 April 2020 — sabse zyada trading! Us din 14.26 Crore shares buy/sell hue!
Ye COVID crash ke baad ka time tha — log panic mein sell kar rahe the ya opportunity dekh ke buy kar rahe the!
Price sirf ₹610 thi — aaj ₹1350+ hai — jo log us din buy kiye unhone 2x se zyada profit kamaya! 😎

In [7]:
# 0 volume wale din dekho
zero_volume = df[df['Volume'] == 0]
print(f"0 volume wale din: {len(zero_volume)}")
print(zero_volume)

0 volume wale din: 2
                   Open         High          Low        Close  Volume  Year  \
Date                                                                           
2025-03-18  1233.921753  1233.921753  1233.921753  1233.921753       0  2025   
2026-01-15  1458.800049  1458.800049  1458.800049  1458.800049       0  2026   

            Month  
Date               
2025-03-18      3  
2026-01-15      1  


In [8]:
# 0 volume wali rows drop karo
df = df[df['Volume'] > 0]

# Confirm karo
print("0 volume rows hataane ke baad shape:", df.shape)
print("0 volume rows remaining:", df[df['Volume'] == 0].shape[0])

0 volume rows hataane ke baad shape: (1547, 7)
0 volume rows remaining: 0


💡 Ek Important Baat:

Ye step humne Day 2 cleaning mein miss kar diya tha — but koi baat nahi! Isliye EDA karna zaroori hota hai — cleaning mein jo cheezein miss ho jaati hain wo EDA mein pakad mein aati hain! Real world mein bhi aisa hi hota hai! 😎

In [9]:
# Daily returns calculate karo
df['Daily_Return'] = df['Close'].pct_change() * 100

print("Daily Returns Statistics:")
print("=========================")
print(f"Average daily return: {df['Daily_Return'].mean():.4f}%")
print(f"Best day return:      {df['Daily_Return'].max():.2f}%")
print(f"Worst day return:     {df['Daily_Return'].min():.2f}%")
print(f"Positive days:        {(df['Daily_Return'] > 0).sum()}")
print(f"Negative days:        {(df['Daily_Return'] < 0).sum()}")

Daily Returns Statistics:
Average daily return: 0.0605%
Best day return:      14.72%
Worst day return:     -13.15%
Positive days:        802
Negative days:        743


💡 Business Insight #5:

802 positive days vs 743 negative days — matlab Reliance zyada din upar rahi neeche se! Ye ek healthy stock ka sign hai! 😎
Aur average 0.06% daily return — ye chhota lagta hai but 365 din mein compound karo toh ~22% annual return banta hai! Ye bahut achha hai!

💡 Business Insight #6:

Best day +14.72% — ye Reliance Rights Issue announcement ka din tha 2020 mein jab company ne bahut badi news di thi! Aur Worst day -13.15% — COVID crash ka time! Ye events price pe kitna bada impact daalte hain! 😮



✅ Day 3 — EDA Complete! 🎊
Aaj tumne ye insights nikalein:
InsightFindingCOVID crash₹395 — sabse kam priceBest year2024 — ₹1414 averageBest monthsJuly-September 📈Worst monthsMarch-April 📉Most active day22 April 2020Fake data2 rows remove kiyeStock growth4x in 6 years!Positive vs Negative802 vs 743 days